Add notJwt, maxTokenExpirySeconds, and skipTypeCheck to CRDs

mstruk opened the pull request:
Signed-off-by: Marko Strukelj <marko.strukelj@gmail.com>

This is a preliminary PR, not yet ready to merge. It is dependent on `strimzi-kafka-oauth` 0.2.0 being released. For the time being it uses 1.0.0-SNAPSHOT version for `strimzi-kafka-oauth` - meaning that it requires `strimzi-kafka-oauth` being built locally from master branch.

Some checklist items may still need to be addressed.

### Type of change

- Enhancement / new feature

### Description

`strimzi-kafka-oauth` module introduced additional oauth authentication config options which need to be exposed through CRDs.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

scholzj commented:
@mstruk Could you update this against the 0.2.0 version of the OAuth library so that we can run the tests?

tombentley commented:
@mstruk is this ready for review yet?

scholzj commented:
@mstruk Culd oyu update this to OAuth 0.2.0 so that it is building? We can than review it and merge it.

mstruk commented:
@scholzj @tombentley Sorry about not responding earlier to this. Bad gmail filter siphoned the notification emails out of my Inbox :O 

I think this PR is ready to merge. I bumped strimzi-kafka-oauth version to 0.2.0 and fixed some docs and tests issues, added some more tests. 

CI doesn't pass but I don't see how the failure would be connected to this PR. I haven't tried manually setup the cluster with the built images for final test, but I don't see why that wouldn't work considering the additional tests pass fine.

scholzj commented:
@mstruk It looks like you didn't updated the CRDs for the Helm Chart:

```
ERROR: Uncommitted changes in derived resources:
M	helm-charts/strimzi-kafka-operator/templates/041-Crd-kafkaconnect.yaml
M	helm-charts/strimzi-kafka-operator/templates/042-Crd-kafkaconnects2i.yaml
M	helm-charts/strimzi-kafka-operator/templates/046-Crd-kafkabridge.yaml
Run the following to add up-to-date resources:
  mvn clean verify -DskipTests -DskipITs \
    && make crd_install \
    && git add install/ helm-charts/ documentation/modules/appendix_crds.adoc cluster-operator/src/main/resources/cluster-roles
    && git commit -s -m 'Update derived resources'
```

Could you please run `make crd_install` and commit the changes?

scholzj commented on api/src/main/java/io/strimzi/api/kafka/model/authentication/KafkaClientAuthenticationOAuth.java:
```diff
@@ -114,4 +116,26 @@ public boolean isDisableTlsHostnameVerification() {
+    @Description("Configure if access token should not be treated as JWT. Set this to `true` if authorization " +
+            "server returns opaque tokens. Defaults to `false`.")
+    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
+    public boolean isNotJwt() {
```
I wonder if we have some better name for this. Maybe just `jwt: true` / `jwt: false`? Ot `jwtFormat`? WDYT @tombentley?

tombentley replied:
Yeah, I wondered about this too. It's always a bit weird having a negative property and having to set notFoo=true. How about `accessTokenIsJwt` defaulting to true?

scholzj replied:
`accessTokenIsJwt` sounds good to me 👍

ppatierno replied:
Yeah tbh I don't like "negative" properties as well :-) `accessTokenIsJwt` sound fine!

mstruk replied:
Very well. Changed to `accessTokenIsJwt`. The same can be changed in strimzi-kafka-oauth library for its next release and the necessary modifications made here behind the scenes.

scholzj commented on api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java:
```diff
@@ -40,6 +40,8 @@
     private String introspectionEndpointUri;
     private String userNameClaim;
+    private boolean skipTypeCheck;
+    private boolean notJwt;
```
Should we set it to `false` as default? Or is it not needed?

mstruk replied:
The default of the 'accessTokenIsJwt' is now `true`. No need for it to be explicitly set.

scholzj commented on api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java:
```diff
@@ -138,6 +140,28 @@ public void setUserNameClaim(String userNameClaim) {
+    @Description("Configure if access token should not be treated as JWT. Set this to `true` if authorization " +
+            "server returns opaque tokens. Defaults to `false`.")
+    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
+    public boolean isNotJwt() {
```
Same as above .... I wonder about the name.

tombentley commented on api/src/main/java/io/strimzi/api/kafka/model/authentication/KafkaClientAuthenticationOAuth.java:
```diff
@@ -114,4 +116,26 @@ public boolean isDisableTlsHostnameVerification() {
     }
+
+    @Description("Set or limit time-to-live of the access tokens to the specified number of seconds. " +
+            "Set this if authorization server returns opaque tokens.")
```
```suggestion
            "This should be set if the authorization server returns opaque tokens.")
```

mstruk replied:
Fixed.

tombentley commented on api/src/main/java/io/strimzi/api/kafka/model/authentication/KafkaClientAuthenticationOAuth.java:
```diff
@@ -114,4 +116,26 @@ public boolean isDisableTlsHostnameVerification() {
+        this.maxTokenExpirySeconds = maxTokenExpirySeconds;
+    }
+
+    @Description("Configure if access token should not be treated as JWT. Set this to `true` if authorization " +
```
```suggestion
    @Description("Configure whether access token should not be treated as JWT. This should be set to `true` if the authorization " +
```

mstruk replied:
Fixed.

tombentley commented on api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java:
```diff
@@ -138,6 +140,28 @@ public void setUserNameClaim(String userNameClaim) {
         this.userNameClaim = userNameClaim;
     }
 
+    @Description("Configure if access token type check should be skipped. Set this to `true` if authorization server " +
```
```suggestion
    @Description("Configure whether the access token type check should be skipped. This should be set to `true` if the authorization server " +
```

mstruk replied:
Fixed.

tombentley commented on api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java:
```diff
@@ -138,6 +140,28 @@ public void setUserNameClaim(String userNameClaim) {
     }
 
+    @Description("Configure if access token type check should be skipped. Set this to `true` if authorization server " +
+            "does not include 'typ' claim in JWT token. Defaults to `false`.")
```
I think typ is not a typo, but I just want to check.

mstruk replied:
Correct. Not a typo.

tombentley commented on api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java:
```diff
@@ -138,6 +140,28 @@ public void setUserNameClaim(String userNameClaim) {
+        this.skipTypeCheck = skipTypeCheck;
+    }
+
+    @Description("Configure if access token should not be treated as JWT. Set this to `true` if authorization " +
```
```suggestion
    @Description("Configure whether access tokens should not be treated as JWT. This should be set to `true` if the authorization " +
```

mstruk replied:
Fixed.

tombentley commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java:
```diff
@@ -1673,6 +1673,26 @@ private static void validateOauth(KafkaListenerAuthenticationOAuth oAuth, String
         }
+
+        if (oAuth.isNotJwt()) {
+            if (oAuth.getJwksEndpointUri() != null) {
```
Should we also check these properties for emptiness, just to be safe?

mstruk replied:
Not following here. The `accessTokenIsJwt` is `boolean` - can't be empty. `jwksEndpointUri` is null unless set in which case it won't be null.

tombentley commented on api/src/main/java/io/strimzi/api/kafka/model/listener/KafkaListenerAuthenticationOAuth.java:
```diff
@@ -138,6 +140,28 @@ public void setUserNameClaim(String userNameClaim) {
+    @Description("Configure if access token type check should be skipped. Set this to `true` if authorization server " +
+            "does not include 'typ' claim in JWT token. Defaults to `false`.")
+    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
+    public boolean isSkipTypeCheck() {
```
Would `checkAccessTokenType` (defaulting to true) be a better name?

mstruk replied:
Sure, why not. Changed the name to `checkAccessTokenType`.

mstruk commented:
We may now want to rename the title of this PR to:

Add accessTokenIsJwt, maxTokenExpirySeconds, and checkAccessTokenType to CRDs

scholzj commented:
@strimzi-ci run tests

scholzj commented:
@strimzi-ci run tests profile=acceptance testcase=io.strimzi.systemtest.KafkaST#testLoadBalancerTls,io.strimzi.systemtest.ConnectST#testKafkaConnectWithFileSinkPlugin
